main.cpp

```diff
@@ -221,17 +242,233 @@ void print_stats(deadline_timer& stats_timer, error_code const& ec)
 	stats_timer.async_wait(std::bind(&print_stats, std::ref(stats_timer), _1));
 }
 
+bool compare_id_prefix(char const* id1, char const* id2)
+{
+	// compare the first 21 bits
+	if (id1[0] != id2[0] || id1[1] != id2[1]) return false;
+	if ((id1[2] & 0xf8) != (id2[2] & 0xf8)) return false;
+	return true;
+}
+
+void generate_id(address const& ip_, uint32_t r, char* id)
```

Comment:
The function got moved because it is now referenced much earlier in the file. The CRC portion of the function has not been changed.